split PublicKey type into context-free type and P2PK type

markblundeberg opened the pull request:
- one for a generic public key without context,
- one for P2PK outputs specifically.
- also clean up the unused TO_ADDRESS_OPS while I'm at it.

Example txn that is P2PK -> P2PK, which is the Satoshi -> Finney txn: `f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16`
(note that input pubkey is not known since we don't know its scriptpubkey)

Along the same lines, we should add a BareMultiSigOutput type eventually. E.g.: `c4aaf7fbec7a9a079e670e50f6a672315451c7618814494ab1f89cf3fd97b3bb`

cculianu commented on lib/address.py:
```diff
@@ -311,24 +308,55 @@ def validate(cls, pubkey):
-        '''Convert to a hexadecimal string.'''
+        '''Convert to a hexadecimal string.
+
+        (do not change this -- many places rely on this to be hexadecimal)
```
🤣

cculianu commented on lib/address.py:
```diff
@@ -311,24 +308,55 @@ def validate(cls, pubkey):
+    @classmethod
+    def from_pubkey(cls, public_key):
+        assert isinstance(public_key, PublicKey)
+        self.public_key = public_key
```
Um. `@classmethod`.. no self...

markblundeberg replied:
yes sir

markblundeberg replied:
I meant to say __init__ I promise!

markblundeberg replied:
wait no that's backwards heh... NamedTuples and their weirdness man

cculianu replied:
I should hope you meant.. `return P2PKOutput(public_key)` .. or somesuch... ;)

cculianu commented on lib/transaction.py:
```diff
@@ -309,7 +309,7 @@ def get_address_from_output_script(_bytes):
     match = [ opcodes.OP_PUSHDATA4, opcodes.OP_CHECKSIG ]
     if match_decoded(decoded, match):
-        return TYPE_PUBKEY, PublicKey.from_pubkey(decoded[0][1])
+        return TYPE_PUBKEY, P2PKOutput.from_pubkey(decoded[0][1])
```
Will this even succeed? Decoded is bytes..no?

markblundeberg commented:
Missed a few places ...

cculianu commented on gui/qt/transaction_dialog.py:
```diff
@@ -351,9 +351,7 @@ def format_amount(amt):
                 cursor.insertText(prevout_hash[-8:] + ":%-4d " % prevout_n, ext)
-                addr = x['address']
-                if isinstance(addr, PublicKey):
-                    addr = addr.toAddress()
```
Yeah this is broken. Ha. In my commit for the asynch inputs dl I also got rid of this thing.  Surprised nothing ever hit this codepath...

cculianu commented:
Hmm.. I'm nervous maybe something was missed.. You were super thorough?

I can also do a painful comb though... using creative grepping...

cculianu commented on ios/ElectronCash/electroncash_gui/ios_native/history.py:
```diff
@@ -347,9 +347,7 @@ def _synch_full(self, wallet):
-                    xa = x['address']
-                    if isinstance(xa, PublicKey):
-                        xa = xa.toAddress()
+                    xa = x.get('address')
```
Lol.. I guess I didn't know what I was doing and just copy-pasta'd here to get it working. *sheepish look*

markblundeberg replied:
Actually everything *should* have 'address' property set right now (at very least to None) because of how parse_sig works. But in future I would like to have that assumption dropped so this is just a bit of forward thinking.

cculianu commented on lib/address.py:
```diff
@@ -311,21 +312,49 @@ def validate(cls, pubkey):
+    def to_ui_string(self):
+        '''Convert to a human-friendly string.
+        '''
+        return 'P2PK(%s)'%(self.pubkey.hex())
```
Are we sure this is safe? Lol.. granted it can't lead to signing errors.. but you'd be surprised with this crazy codebase. Ha ha.

markblundeberg replied:
The idea is that `to_ui_string` here is only going to be called in same places where it's called for ScriptOutputs, so that it will be totally ok. :-)

But, I must continue digging. :-)

cculianu replied:
You are literally adding a P2PKOutput to `inputs` in this very PR -- which will end up causing the same thing that broke in 4.0.4 release... because code relies on `to_ui_string` everywhere.. :)

cculianu commented on lib/address.py:
```diff
@@ -342,15 +371,9 @@ def to_scripthash_hex(self):
-    def __str__(self):
-        return self.to_ui_string()
-
+    # no __str__ method
```
Here is where I'm nervous.. hehe.  I'll try and see if anything expects the PublicKey output to have a str(e) -> hex bytes.  Hopefully not.

cculianu commented:
Regarding the bare multisig -- ha ha! I rather like seeing that whole wordy scriptoutput. :)  But yeah -- I guess you'd want it to collapse down to something more succinct UI-wise?

cculianu commented on lib/wallet.py:
```diff
@@ -90,7 +90,7 @@ def append_utxos_to_inputs(inputs, network, pubkey, txin_type, imax):
         address = Address.from_pubkey(pubkey)
     else:
-        address = PublicKey.from_pubkey(pubkey)
+        address = P2PKOutput.from_pubkey(pubkey)
```
This will break things -- no? P2PKOutput as INPUT will lead to that same bug as before.  Please fix this if that's the case.

markblundeberg replied:
This is intentional -- note that the following line `address.to_scripthash_hex()` will not work on PublicKey anymore (which is exactly my intent -- generic public keys do not have a script!).

markblundeberg replied:
Still, I want to test these things fully so no merge yet :)

cculianu replied:
Ok but then go down to like 99.  We have already seen `input['address']`   where `to_ui_string` is not either an address or hex bytes fail in some of the wallet code.

cculianu commented:
So Mark, I feel this PR still carries with it some risk.  I don't see the huge advantage of splitting it into two types.  What's the end-goal or justification here?

if this were C++ I would normally be all-for it -- as it would buy us additional type safety.

But this being python where types are completely runtime-determined doesn't buy us much on that front, at least.

What is the thing I'm missing?

In my mind all I see this is somehow breaking again because `PublicKey` is implicitly possibly used in places we missed.

It also means now you have to translate PublicKey -> P2PKOutput in code that "copies" inputs -> outputs.

Conceptually for idiots it's also easier to just have 3 types to worry about: PublicKey, Address, and Script.  Now we have 4 -- and one of them only if output.

Not sure what is happening.  I'm still worried/traumatized/have PTSD from Python. :)

Do argue your point if you can...

markblundeberg commented:
> It also means now you have to translate PublicKey -> P2PKOutput in code that "copies" inputs -> outputs.

The idea is that the "address"es in both inputs and outputs should never be PublicKey. If I've missed a spot, let me know. :-) In fact I've removed the to_script() method from PublicKey so it should be literally impossible to appear in either inputs or outputs.

Hmm maybe P2PKOutput was a poor name, it's not just meant for outputs.

Anyway, maybe a picture helps. Here's the satoshi -> finney transaction after I've rebased onto your tx input fetcher. :-)

![Screenshot from 2019-05-31 08-47-03](https://user-images.githubusercontent.com/36528214/58717738-c410d200-8380-11e9-9e9d-ced6b4ea1392.png)

cculianu commented:
I did let you know -- see my comment about breakage above... :)

Hmm. I am just not sure -- what is the "win" here, exactly?  Just that nice UI? We can do that *in* the UI without touching a class that is raped and abused everywhere.

I don't get what the win is -- aside from surprise breakages 3 weeks from now because Python lacks type safety...?

cculianu commented:
Please if you have time provide a justification or explanation for this change -- I tried really carefully to read what you have written thus far and it's not clear to me what your motivation was/is.  Again I ask: that is the benefit to this refactoring?

Thank you.

markblundeberg commented:
> Please if you have time provide a justification or explanation for this change -- I tried really carefully to read what you have written thus far and it's not clear to me what your motivation was/is. Again I ask: what is the benefit to this refactoring?
> 
> Thank you.

Ah, my motivation is to clean up the ambiguity that led us to the to_ui_string bug in the first place.

PublicKey currently serves dual duty:
1. Representing public keys in a generic context that is independent of script type: it gets used for all public keys in keystore (be they for p2pkh or multisig), and the validate method gets used to validate encoding of public keys for P2PKH, etc. When you open Address Details dialog it shows you a PublicKey object for the pubkeys list, be it P2PKH or multisig.
2. It gets used as an 'address type' analogous to Address and ScriptOutput. In this role it represents a very specific type of script: pay-to-pubkey. In this role, to_ui_string is used to generate the text for the Transaction details dialog.

The reason I had the bug with `to_ui_string` is that I was attempting to change behaviours relating to the second role (address), where already we know from the example of ScriptOutput that you can put anything you want in `to_ui_string`. But that had unintended impact due to the overloading of PublicKey with its first role.

This PR removes the 'address-ness' from PublicKey, so now it is only a generic public key meant for keystore and validation. All the 'address-ness' is moved into a new class P2PKOutput, which is a sibling of ScriptOutput and Address.

cculianu commented:
Thanks for the explanation -- it's late now so it didn't all sink in yet. I'll re-read it again tomorrow when I'm fresh. 

I kept mentioning this: but there is still code that relies on the "Addressness" to generate 'x_pubkeys' for the wallet using the famed `to_ui_string` (specifically, if you look at the ImportedAddress Wallet in wallet.py).

Can you fix that code too then? That's what broke last time and it will break again with this PR right now as it stands.  It will break anytime you put this 'P2PKOutput' class into an input's 'address' field, as was done in this very PR!

markblundeberg commented:
> Thanks for the explanation -- it's late now so it didn't all sink in yet. I'll re-read it again tomorrow when I'm fresh.
> 
> I kept mentioning this: but there is still code that relies on the "Addressness" to generate 'x_pubkeys' for the wallet using the famed `to_ui_string` (specifically, if you look at the ImportedAddress Wallet in wallet.py).
> 
> Can you fix that code too then? That's what broke last time and it will break again with this PR right now as it stands. It will break anytime you put this 'P2PKOutput' class into an input's 'address' field, as was done in this very PR!

Hmm I don't see how P2PKOutput in 'address' field will ever hurt anything. If so then the txdetails updater would also damage things when it sticks a ScriptOutput into the 'address' field of an input, no?

You mean this code right? https://github.com/Electron-Cash/Electron-Cash/blob/bfe3a23d678b6d1dc9cfded08276c87eadf5f0b4/lib/wallet.py#L2004-L2009

In this case it's definitely not a P2PKOutput (wouldn't even make sense since we don't have P2PK wallets). Since it's gotten from keystore, `pubkey` here is always going to be a PublicKey object for which `to_ui_string` is just hex.

Just to make sure I'm not crazy, I checked and yes it does work. :)

cculianu commented:
Oh yeah you're right that code path will always generate its own `PublicKey` from `Address`, and `txin['address']` will always be `p2pkh`. Sorry, my bad. Heh

TinosNitso commented:
2022 and EC still can't detect when someone uses P2PK to pay our wallet address. The payment shows up on Blockchain.com, but not in our wallet. [Here](https://www.blockchain.com/bch/tx/a1099a8094d14fe9bd6fe970f29a1a1064e0c44bb77882903a6e2a0cc6a9c4e3)'s an example of a 151B Schnorr P2PK I did using the Console. The trick was to sign the input's (Preimage) scriptCode as `<PubKey> OP_CHECKSIG`. 

Another example is if you add Block 10's coinbase [address](https://www.blockchain.com/bch/address/qqmgj6e8nrhwcr7q8j2m2234w3yw6r0agczxy2pk5m) to a watching-only wallet, you'll see the EC balance is just 0.0001 BCH. So it's impossible for Satoshi to detect many of his own coins using EC!

I'm interested in coding a covenant (in [AutoCove](https://github.com/TinosNitso/AutoCove-Plugin)) to auto-return to P2PK senders, but without EC support it seems a bit too tedious to test. It could be a bit embarrassing or time consuming if someone sends to an important address using a P2PK output.
